### DISCUSSION TITLE
Integrate whisper-turbo model using speechbrain

### DISCUSSION THREAD

**Discussion by wredan** [speechbrain/speechbrain#2758] [Q&A]
Hi everyone,

I would like to implement and fine-tune the Whisper Turbo model released by OpenAI on HF: https://huggingface.co/openai/whisper-large-v3-turbo. I have been reading the SpeechBrain documentation for a while now. Could you point me to the right resources or documentation to follow in order to integrate this model with SpeechBrain?

Do I need to implement it through the Brain class and similar methods?

Thank you in advance!

**Comment by TParcollet** [comment]
Hi, have a look at our whisper fine-tuning recipes in Librispeech ASR.

**Comment by wredan** [comment]
Hi, thanks for your time.

I’m looking at the CommonVoice recipe actually and tried to adapt the code to use it with Whisper Turbo. I think the main issue resides in the Whisper Turbo HF repo: https://huggingface.co/openai/whisper-large-v3-turbo

They do not provide any .bin or checkpoint, so I get the following error:

```bash
FileNotFoundError: openai/whisper-large-v3-turbo does not contain a .bin or .ckpt checkpoint !
```
Is there any way to avoid this issue? I found a shell script to download the .bin of the model here: https://github.com/ggerganov/whisper.cpp/blob/master/models/download-ggml-model.sh

Is there any workaround that would allow me to use SpeechBrain for fine-tuning and inference with this model?

Thank you

**Reply by Adel-Moumen** [reply]
Hey @wredan, can you please fetch the last commits in the dev branch? We merged a fix in https://github.com/speechbrain/speechbrain/pull/2763 that should allows you to use this model :)

**Reply by wredan** [reply]
Hi, awesome team! :D

I fetched the fix, and now the recipe code is working :)

I’m able to generate a .ckpt file to use with inference for `WhisperASR`. I tested it with the usual `example-it.wav` file provided for the Italian test with SpeechBrain. The reference content, I assume, is:

```
BUONGIORNO A TUTTI E BENVENUTI A BORDO
```

However, what I get using my generated turbo checkpoint with the following code:

```py
result = asr_model.transcribe_file(audio_path)
print(" ".join(segment.words for segment in result))
```
by the way is:
```
BUONG GIORNO A TUTTI E BENVENU
```
The turbo model inference output using the official repo python code gives me the following result, to compare:
```
Buongiorno a tutti e benvenuti a bordo!
```
which is the most accurate result, according to WER.

Am I missing something? I’m aware I need to set up a text processor pipeline for the output of `transcribe_file` (e.g., for punctuation, etc.). And if you could point me to some resources for that, I’d be really grateful!

However, I was expecting a better transcription from this model. I also tested it with a 1-minute-long audio file, but it didn’t perform well, truncating after the first few seconds of the batch.

Thank you all for the support!

**Reply by Adel-Moumen** [reply]
What is the decoding function? Greedy search?

**Reply by wredan** [reply]
Yes, here the yaml I'm using, adapted it from the recipe:

```yaml
# Feature parameters
sample_rate: 16000

# Model parameters
freeze_whisper: True
freeze_encoder: True
encoder_only:  False                  # set True if you want feature extraction
whisper_hub: openai/whisper-large-v3-turbo

# Path to che model checkpoint
whisper_ckpt_path: 
  ./training/whisper/large-v3-turbo/save/CKPT+2024-11-25+11-03-34+00/whisper.ckpt
# Path where the "whisper_hub" will be saved
save_path: 
  ./local_models/whisper_large_v3_turbo/hub_model
language: it
task: transcribe

# Decoding parameters
min_decode_ratio: 0.0
max_decode_ratio: 1.0
test_beam_size: 5

# Normalize inputs with the same normalization done in the paper (https://cdn.openai.com/papers/whisper.pdf). Refer to Appendix C for further information.
normalized_transcripts: True

# Data files

# Not required in inference

############################## Models ##########################################

whisper: !new:speechbrain.lobes.models.huggingface_transformers.whisper.Whisper
  source: !ref <whisper_hub>
  freeze: !ref <freeze_whisper>
  freeze_encoder: !ref <freeze_encoder>
  encoder_only: !ref <encoder_only>
  save_path: !ref <save_path>
  language: !ref <language>
  task: !ref <task>

############################## Final decoding (NOT the Speech Recognizer whisper decoder) ##########################################

# https://speechbrain.readthedocs.io/en/latest/tutorials/tasks/speech-recognition-from-scratch.html#overview-of-speech-recognition
# Choose one
# - greedy_search: simpler ed efficient
# - beam_search: accurated but not so efficient

decoder: !new:speechbrain.decoders.seq2seq.S2SWhisperGreedySearcher
  model: !ref <whisper>
  min_decode_ratio: !ref <min_decode_ratio>
  max_decode_ratio: !ref <max_decode_ratio>

# decoder: !new:speechbrain.decoders.seq2seq.S2SWhisperBeamSearcher
#   module: !ref <whisper>
#   min_decode_ratio: !ref <min_decode_ratio>
#   max_decode_ratio: !ref <max_decode_ratio>
#   beam_size: !ref <test_beam_size>

modules:
  whisper: !ref <whisper>
  decoder:  !ref <decoder>

pretrainer: !new:speechbrain.utils.parameter_transfer.Pretrainer
    loadables:
        whisper: !ref <whisper>
    paths:
      whisper: !ref <whisper_ckpt_path>
```

I did test the `S2SWhisperBeamSearcher` but I get the following error log:

```bash
Traceback (most recent call last):
  File "path_to_script/test_pretrained_it_models.py", line 175, in <module>
    main()
  File "path_to_script/test_pretrained_it_models.py", line 169, in main
    transcriptions, times = transcribe_audio_files(audio_files, model_name, model_info, device, error_log_path)
                            ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "path_to_script/test_pretrained_it_models.py", line 105, in transcribe_audio_files
    asr_model = load_model(model_name, model_info, device)
                ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "path_to_script/test_pretrained_it_models.py", line 68, in load_model
    return model_info["class"].from_hparams(
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "site-packages/speechbrain/inference/interfaces.py", line 508, in from_hparams
    hparams = load_hyperpyyaml(
              ^^^^^^^^^^^^^^^^^
  File "site-packages/hyperpyyaml/core.py", line 188, in load_hyperpyyaml
    hparams = yaml.load(yaml_stream, Loader=loader)
              ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "site-packages/yaml/__init__.py", line 81, in load
    return loader.get_single_data()
           ^^^^^^^^^^^^^^^^^^^^^^^^
  File "site-packages/ruamel/yaml/constructor.py", line 116, in get_single_data
    return self.construct_document(node)
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "site-packages/ruamel/yaml/constructor.py", line 120, in construct_document
    data = self.construct_object(node)
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "site-packages/ruamel/yaml/constructor.py", line 147, in construct_object
    data = self.construct_non_recursive_object(node)
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "site-packages/ruamel/yaml/constructor.py", line 188, in construct_non_recursive_object
    for _dummy in generator:
  File "site-packages/ruamel/yaml/constructor.py", line 633, in construct_yaml_map
    value = self.construct_mapping(node)
            ^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "site-packages/ruamel/yaml/constructor.py", line 429, in construct_mapping
    return BaseConstructor.construct_mapping(self, node, deep=deep)
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "site-packages/ruamel/yaml/constructor.py", line 244, in construct_mapping
    value = self.construct_object(value_node, deep=deep)
            ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "site-packages/ruamel/yaml/constructor.py", line 147, in construct_object
    data = self.construct_non_recursive_object(node)
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "site-packages/ruamel/yaml/constructor.py", line 183, in construct_non_recursive_object
    data = constructor(self, tag_suffix, node)
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "site-packages/hyperpyyaml/core.py", line 481, in _construct_object
    return callable_(*args, **kwargs)
           ^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "site-packages/speechbrain/decoders/seq2seq.py", line 1904, in __init__
    bos_index=module[0].bos,
              ~~~~~~^^^
TypeError: ('Invalid argument to class speechbrain.decoders.seq2seq.S2SWhisperBeamSearcher', "'Whisper' object is not subscriptable")
```

**Reply by wredan** [reply]
Well, the issue with the `S2SWhisperBeamSearcher` could be because I saved the checkpoint with greedy search during fine tuning, now that I think about it. Is that right?

**Reply by Adel-Moumen** [reply]
And what about the official python impl? Beam search?

**Reply by Adel-Moumen** [reply]
> Well, the issue with the `S2SWhisperBeamSearcher` could be because I saved the checkpoint with greedy search during fine tuning, now that I think about it. Is that right?

No no, there’s something happening. Can you ping me the official impl pls ?

**Reply by wredan** [reply]
yes sure, here is their implementation: https://github.com/openai/whisper/blob/main/whisper/model.py

**Reply by Adel-Moumen** [reply]
Ty. Also make sure that the colab also includes the audio file so that i can debug locally thanks

**Reply by wredan** [reply]
Actually the interesting part should be at: https://github.com/openai/whisper/blob/main/whisper/decoding.py

Anyway do you mean I should create the colab with my current implementation? Do you need the inference or the fine tuning code I'm using? Both?

Thanks for the support

**Reply by Adel-Moumen** [reply]
Just the inference with the audio file that you are using

**Reply by wredan** [reply]
Hi, here is the colab: https://colab.research.google.com/drive/1SCg-Vz7XM5lqE8BWcdlT1GOFuy7zDk-M?usp=sharing

Thank you again, I hope there is something with my code or I'm glad to help with speechbrain :)

**Reply by ZirwaZafar** [reply]
Great 👍 work

**Reply by Adel-Moumen** [reply]
> Hi, here is the colab: [colab.research.google.com/drive/1SCg-Vz7XM5lqE8BWcdlT1GOFuy7zDk-M?usp=sharing](https://colab.research.google.com/drive/1SCg-Vz7XM5lqE8BWcdlT1GOFuy7zDk-M?usp=sharing)
> 
> Thank you again, I hope there is something with my code or I'm glad to help with speechbrain :)

I am currently trying to reproduce your issue. Currently, I am trying the inference and it seems to match with OA results (taking the same model, the output we have is the same). I believe the issue is really located at the training stage. So I am in the process of re-training a Whisper model.

**Comment by Adel-Moumen** [comment]
Hey, FYI, I am going to work on this issue during the weekend! :)
